biz.aQute.resolve/src/biz/aQute/resolve/MissingRequirementsResource.java

```diff
@@ -0,0 +1,107 @@
+package biz.aQute.resolve;
+
+import static org.osgi.framework.namespace.IdentityNamespace.IDENTITY_NAMESPACE;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.osgi.framework.namespace.PackageNamespace;
+import org.osgi.resource.Capability;
+import org.osgi.resource.Requirement;
+import org.osgi.resource.Resource;
+
+import aQute.bnd.osgi.resource.CapReqBuilder;
+import aQute.lib.exceptions.Exceptions;
+
+public class MissingRequirementsResource implements Resource {
+
+	public static final String	MISSING_REQUIREMENT	= "bnd.missing.requirement";
+
+	List<Capability> allCapabilities = new CopyOnWriteArrayList<>();
+
+	public MissingRequirementsResource() {
+		CapReqBuilder builder = new CapReqBuilder(IDENTITY_NAMESPACE);
+
+		try {
+			builder.setResource(this);
+			builder.addAttribute(IDENTITY_NAMESPACE, MISSING_REQUIREMENT);
+		} catch (Exception e) {
+			Exceptions.duck(e);
+		}
+
+		allCapabilities.add(builder.buildCapability());
+	}
+
+	public List<Capability> addMissingRequirement(Requirement requirement) {
+		if (isOptional(requirement)) {
+			return new ArrayList<>();
+		}
+
+		CapReqBuilder builder = new CapReqBuilder(MISSING_REQUIREMENT);
+		builder.setResource(this);
+
+		try {
+			builder.addAttribute(MISSING_REQUIREMENT, requirement);
+		} catch (Exception e) {
+			Exceptions.duck(e);
+		}
+
+		Capability capability = builder.buildCapability();
+		allCapabilities.add(capability);
```

Comment:
The resolver will throw an exception of empty list providers is returned for a mandatory requirement! We cannot collect all missing requirements if we've cause an exception to be thrown from the resolver. So we must return something. That something should be a Capability. It seemed logical to link all missing Capabilities to a single resource.